pkg/kv/kvserver/logstore/logstore.go

```diff
@@ -540,16 +541,56 @@ func LoadEntries(
 		n = 100
 	}
 	ents := make([]raftpb.Entry, 0, n)
+	// TODO(pav-kv): make all the entry getting limits soft. In most cases we
+	// already paid the cost of fetching an extra entry, so we should rather
+	// return it. This would also simplify the code below that takes both the
+	// dynamic memory budgeting and the static limit into account.
+	//
+	// TODO(pav-kv): encapsulate the maxBytes + dynamic limit in a type, pass it
+	// down the stack.
+	ents, _, hitIndex, _ := eCache.Scan(ents, rangeID, lo, hi, maxBytes)
+
+	var totalSize uint64 // tracks total size of the returned entries
+	// exceededMaxBytes is set to true when the last or next-to-last entry brings
+	// the total size over maxBytes, or the memory budget limit is reached. Entry
+	// iteration stops when this is true.
+	exceededMaxBytes := false
+	// budgetFor is called sequentially on each entry. We return a prefix of
+	// entries for which budgetFor() returned true, and stop iteration as soon as
+	// sets exceededMaxBytes to true.
+	budgetFor := func(size uint64) bool {
```

Comment:
This can be encapsulated into a type `{totalSize, maxBytes, *account}`, which then can be used by all parts of this stack in a more stateful way:

- `eCache.Scan` will accumulate the right amount of entries
- if `!exceededMaxBytes`, this type is further used by the code below to add some entries from the storage
- if still `!exceededMaxBytes`, raft will add some more entries from `unstable`.

Captured by the TODO above.

Reply:
Added a `sizeHelper` type and covered it with tests.